### DISCUSSION TITLE
SPDX output format and vuln scanner

### DISCUSSION THREAD

**Discussion by goneall** [aquasecurity/trivy#6308] [Ideas]
### Question

I noticed that when I try to run trivy with `--output spdx-json` and `--scanners vuln` I get the following message:

```
"--format spdx" and "--format spdx-json" disable security scanning
```

A couple of questions:
- Is vuln scanning actually disabled as the message suggests?
- What is the reason for disabling vuln scanning when outputting to SPDX?

I'm relatively new to Trivy and haven't had much experience with the vuln scanner, but I was thinking vulnerabilities could be reported as external references per the [how to use annex](https://spdx.github.io/spdx-spec/v2.3/how-to-use/).

I can move this over to the new feature discussion if it is more appropriate.

I'm not a Go programmer, so I can't help much with the implementation, but I can provide help on the SPDX format.

### Target

None

### Scanner

Vulnerability

### Output Format

SPDX

### Mode

Standalone

### Operating System

Docker image

### Version

```bash
Version: 0.49.1
Vulnerability DB:
  Version: 2
  UpdatedAt: 2024-03-12 18:11:44.883502428 +0000 UTC
  NextUpdate: 2024-03-13 00:11:44.883502148 +0000 UTC
  DownloadedAt: 2024-03-12 19:09:55.2745624 +0000 UTC
```
```

**Comment by DmitriyLewen** [comment]
Hello @goneall 
Thanks for your interest to Trivy!

> Is vuln scanning actually disabled as the message suggests?

You are right. Trivy only detects packages and inserts them into SPDX format (without vulnerability checking).

> What is the reason for disabling vuln scanning when outputting to SPDX?

In version 2.2 there were no instructions for inserting advisories into the SPDX format.
And it looks like we missed [Annex K](https://spdx.github.io/spdx-spec/v2.3/how-to-use) after updating to 2.3

`externalRefs` doesn't support parts of information (e.g. we can't insert fixed version, severity).
But it is better than nothing.
@knqyf263 wdhy? I think we can think about using `externalRefs` for advisories (as docs say - https://spdx.github.io/spdx-spec/v2.3/how-to-use/#k11-linking-to-an-advisory)

**Reply by goneall** [reply]
Thanks @DmitriyLewen for the quick reply.

Adding the advisories would be a big help.  I'm working on a project to generate SPDX SBOMs for open source projects which include references to known vulnerabilities and license information - Trivy comes very close to meeting the needs for the project.

SPDX 3.0 will support the additional fields, but there it a lot of value adding the advisory information to SPDX 2.X even if it is not complete.

**Reply by goneall** [reply]
Just checking to see if there are any updates on this request - we're in the process of selecting an app to use for generating SBOMs for mission critical projects at the Linux Foundation and would like to use Trivy if it supports the features we need.

**Reply by knqyf263** [reply]
Sorry I missed this.

@DmitriyLewen There are two options:

1. Adding vulnerabilities to SPDX 2.3 Trivy currently supports
2. Moving to SPDX 3.0 and adding vulnerabilities

If v3.0 is easy to migrate, I'd go with option 2.

**Reply by knqyf263** [reply]
@goneall I was looking into [this repository](https://github.com/spdx/spdx-examples), but didn't find so much examples for SPDX 3.0 other than [this one](https://github.com/spdx/spdx-examples/blob/7b3c5a215d6acd43ef42b2261f7decfbf8aec879/software/example13/spdx-3.0/example-13-spdx-3.json#L2), which doesn't contain vulnerabilities. Can we find any examples somewhere else?

**Reply by DmitriyLewen** [reply]
> If v3.0 is easy to migrate, I'd go with option 2.

[scheme](https://spdx.github.io/spdx-spec/v3.0/model/schema.json) has been greatly changed.
Also `github.com/spdx/tools-golang` team is still thinking about a better 3.0 implementation - https://github.com/spdx/tools-golang/pull/240.

Looks like migration to 3.0 will take quite some time
So I think we are forced to choose option 1.

**Reply by goneall** [reply]
@knqyf263 We're actively working on adding SPDX 3.0 examples to the examples repo, in the interim, you can find examples in the spec itself for vulnerabilities - there have been some recent updates for the upcoming 3.0.1 patch release, so I would suggest looking at the in development [SPDX Spec 3 model security profile](https://github.com/spdx/spdx-3-model/tree/main/model/Security)

**Reply by knqyf263** [reply]
OK. SPDX 3.0 seems to be too cutting-edge. Let's add vulnerabilities to v2.3.
@DmitriyLewen Can you please convert it to an issue?

**Reply by DmitriyLewen** [reply]
#7211

**Comment by goneall** [comment]
Agree with @DmitriyLewen - SPDX 3 is a significant change and the golang tools are still being updated.  For the Linux Foundation project, we're targeting SPDX 2.3 in the short term then updating to 3.0 after the 3.0.1 patch release.

**Reply by knqyf263** [reply]
Thanks for your input. We'd go for SPDX 2.3 for now.

**Comment by goneall** [comment]
Thanks @knqyf263 and @DmitriyLewen  for the quick responses and follow-up questions :)

**Comment by DmitriyLewen** [comment]
Track #7211
